### DISCUSSION TITLE
sidekiq-ent uniqueness is still locked after Sidekiq::Worker.drain_all

### DISCUSSION THREAD

**Discussion by jarl-dk** [sidekiq/sidekiq#6309] [Q&A] [answered]
According to the docs https://github.com/sidekiq/sidekiq/wiki/Ent-Unique-Jobs:
> a second job can be pushed to Redis after 10 minutes or after the first job has successfully processed

I have a `MyJob` with `sidekiq_options unique_for: 9999.years`. I have a test where I `MyJob.perform` a job, then `Sidekiq::Worker.drain_all` and verify the behaviour. The second time I run this test the `MyJob.perform` prints out
```
Skipping enqueue for MyJob, not unique (lock held by JID 1b79982302dc2e8ffc1e202b)
```
So somehow the first tests locks this (as expected), but `Sidekiq::Worker.drain_all` does not release this lock when successfully processed (as documentation says).

Is there a bug? or am I doing something wrong? or do I misunderstand the documentation?

**Comment by mperham** [comment]
In testing mode, you'd need to add any required server middleware to Sidekiq::Testing.server_middleware for it to run when draining.

**Reply by jarl-dk** [reply]
What server middleware is required in this case? And how do I add that to `Sidekiq::Testing.server_middleware`?

**Reply by mperham** [reply]
I don't recommend using uniqueness in testing, it can cause test cases to flap if you aren't careful to reset your Redis database between tests because it effectively creates global locks in Redis. That said:

https://github.com/sidekiq/sidekiq/wiki/Testing#testing-server-middleware

If you run `bundle open sidekiq-ent`, the uniqueness server middleware is loaded at the bottom of `lib/sidekiq-ent/unique.rb`. You can create a similar Sidekiq::Testing.server_middleware block in your `{spec,test}/helper.rb`.

**Reply by mperham** [reply, marked as answer]
```ruby
Sidekiq::Testing.server_middleware do |chain|
  chain.add Sidekiq::Enterprise::Unique::Server
end
```

**Reply by jarl-dk** [reply]
Thanks
